OpenRA.Game/Primitives/LongBitSet.cs

```diff
@@ -0,0 +1,175 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace OpenRA.Primitives
+{
+	static class LongBitSetAllocator<T> where T : class
+	{
+		static readonly Cache<string, long> Bits = new Cache<string, long>(Allocate);
+		static long nextBits = 1;
+
+		static long Allocate(string value)
+		{
+			lock (Bits)
+			{
+				var bits = nextBits;
+				nextBits <<= 1;
```

Comment:
This will silently break (wrap to 0 and then stay there) when we run out of bits. Can we please check for `nextBits == 0` after this and throw a `System.OverflowException`?